### ISSUE TITLE
Full minification failing error

### ISSUE THREAD

**Issue by stolliemods** [malforge/mdk2#50]
Hi,

I'm not sure what information I need to provide, I'm a pretty amateur coder. When I use Lite the script minifies just fine. 

When I try to Full I get the below error. I've tried using all Explicit types in the script and removing any mention of 'var' in the entire script and that error still shows up.

Is there anything else I can look at to see what's going wrong?

1>Starting symbol renaming...
1>Symbol renaming complete in 00:00:00.8662093
1>Writing intermediate symbol map to C:\Users\cheyn\source\repos\AutoDockingUpdatedMDK\obj\symbol.map
1>Compilation failed with errors.
1>The contextual keyword 'var' may only appear within a local variable declaration or in script code
1>'object' does not contain a definition for 'Value' and no accessible extension method 'Value' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)
1>'object' does not contain a definition for 'Value' and no accessible extension method 'Value' accepting a first argument of type 'object' could be found (are you missing a using directive or an assembly reference?)
1>Implicitly-typed variables must be initialized
1>Implicitly-typed variables must be initialized
1>C:\Users\cheyn\.nuget\packages\mal.mdk2.pbpackager\2.0.13\build\Mal.Mdk2.PbPackager.props(18,9): error MSB3073: The command ""C:\Users\cheyn\.nuget\packages\mal.mdk2.pbpackager\2.0.13\build\..\tools\mdk.exe" pack "C:\Users\cheyn\source\repos\AutoDockingUpdatedMDK\AutoDockingUpdatedMDK.csproj"  -configuration Debug -interactive" exited with code -1.
1>Done building project "AutoDockingUpdatedMDK.csproj" -- FAILED.

**Comment by malware-dev** [comment]
First; make sure you have everything up to date ([see wiki](https://github.com/malforge/mdk2/wiki)). If the error still occurs, you can try to enable "trace" mode in the .ini file to get more information, and a whole bunch of intermediate code files. Please try to narrow down where the error actually occurs.

**Comment by stolliemods** [comment]
Thanks, I've confirmed I needed to update the MDK2 Templates. Nuget packages is up to date, I did that earlier today and confirmed no further update pending as I write this.

I turned on trace mode previously which is where that above log information came from. I've had a look through the intermediate files and can't see where it's going wrong. The only things I see are: -

1. The WhitespaceTrimmer intermediate file looks like it worked.
2. TypeSorter seems to have worked, the Classes are in alphabetical order.
3. Symbol.map appears to be populated with a symbol against everything.
4. It seems to be failing at the point of mapping those symbols to the intermediate script.

Is there any way for me to see specfically which symbol it fails at, see any more debugging info or check which type references should not be changed and if they are in the symbol map?

**Comment by malware-dev** [comment]
No, the way this usually is done is to try to see where it fails in the processed code vs where that is in the unprocessed code. It would not be in the symbol map as it's not _renaming_ anything in this case, simply trying to use `var` instead of the type itself, to shorten it. 

It's directly related to a specific way _you_ write your code, which is why it needs to be narrowed down so I can fix it.

**Comment by malware-dev** [comment]
I will help you if that's the only option, but I beg you to try for yourself first. I don't have much free time to work on this project, so the more you guys help me out, the better for all of us 🙏

**Comment by stolliemods** [comment]
All good mate, dont stress, I'll try and narrow it down to where its happening, thanks for your help!

Worse case scenario I'll have to refactor and use less descriptive method and variable names 😄

**Comment by malware-dev** [comment]
I'd prefer to be able to fix the problem than you having to refactor, since someone else might be getting the issue and there's already two issues I have no good solutions for :P

**Comment by stolliemods** [comment]
The joys of coding 😄 

So I've narrowed down one of them by the good old process of turrning things off until it works haha

I can't figure out why it doesn't like it though, it spits the dummy when at the localDireciton variable in this class: -

`public class ThrusterGroup
{
    private readonly Program parent_program;
    public double[] ANS;
    public IMyTerminalBlock directionReferenceBlock;
    public int directionSign = 1;
    public ThrusterGroup[] finalThrusterGroups;
    public Vector3D finalThrustForces;

    // Temporary variables, defined here so they aren't defined every frame.
    public double lambdaResult;
    public Base6Directions.Direction localThrustDirection;
    public double[,] matrixM;


    /// <summary>Force In Newtons</summary>
    public double MaxThrust;

    public List<IMyThrust> thrusters;
    public Vector3D worldThrustDirection;

    public ThrusterGroup(Program _parent_program, Base6Directions.Direction direction,
        IMyTerminalBlock _directionReferenceBlock)
    {
        parent_program = _parent_program;
        directionReferenceBlock = _directionReferenceBlock;
        thrusters = new List<IMyThrust>();
        MaxThrust = 0;
        //LocalThrustDirection = direction;
        UpdateWorldDirection();
        ANS = new double[3];
        finalThrustForces = new Vector3D();
        finalThrusterGroups = new ThrusterGroup[6];
        matrixM = new double[3, 3];

        if (localThrustDirection == Base6Directions.Direction.Down ||
            localThrustDirection == Base6Directions.Direction.Right ||
            localThrustDirection == Base6Directions.Direction.Backward)
            directionSign = -1;
        else
            directionSign = 1;
    }

    public void AddThruster(IMyThrust thruster)
    {
        thrusters.Add(thruster);
        MaxThrust += thruster.MaxEffectiveThrust;
        thruster.CustomName = localThrustDirection.ToString() + " " + thrusters.Count.ToString();
    }

    public void UpdateWorldDirection()
    {
        worldThrustDirection = directionReferenceBlock.WorldMatrix.GetDirectionVector(localThrustDirection);
    }
}`

**Comment by malware-dev** [comment]
That should do it. This is small enough, I'll take it from here. Thank you!On Oct 22, 2024 23:44, Stollie ***@***.***> wrote:
The joys of coding 😄
So I've narrowed down one of them by the good old process of turrning things off until it works haha
I can't figure out why it doesn't like it though, it spits the dummy when at the localDireciton variable in this class: -
`public class ThrusterGroup
{
private readonly Program parent_program;
public double[] ANS;
public IMyTerminalBlock directionReferenceBlock;
public int directionSign = 1;
public ThrusterGroup[] finalThrusterGroups;
public Vector3D finalThrustForces;
// Temporary variables, defined here so they aren't defined every frame.
public double lambdaResult;
public Base6Directions.Direction localThrustDirection;
public double[,] matrixM;


/// <summary>Force In Newtons</summary>
public double MaxThrust;

public List<IMyThrust> thrusters;
public Vector3D worldThrustDirection;

public ThrusterGroup(Program _parent_program, Base6Directions.Direction direction,
    IMyTerminalBlock _directionReferenceBlock)
{
    parent_program = _parent_program;
    directionReferenceBlock = _directionReferenceBlock;
    thrusters = new List<IMyThrust>();
    MaxThrust = 0;
    //LocalThrustDirection = direction;
    UpdateWorldDirection();
    ANS = new double[3];
    finalThrustForces = new Vector3D();
    finalThrusterGroups = new ThrusterGroup[6];
    matrixM = new double[3, 3];

    if (localThrustDirection == Base6Directions.Direction.Down ||
        localThrustDirection == Base6Directions.Direction.Right ||
        localThrustDirection == Base6Directions.Direction.Backward)
        directionSign = -1;
    else
        directionSign = 1;
}

public void AddThruster(IMyThrust thruster)
{
    thrusters.Add(thruster);
    MaxThrust += thruster.MaxEffectiveThrust;
    thruster.CustomName = localThrustDirection.ToString() + " " + thrusters.Count.ToString();
}

public void UpdateWorldDirection()
{
    worldThrustDirection = directionReferenceBlock.WorldMatrix.GetDirectionVector(localThrustDirection);
}

}`

**Comment by stolliemods** [comment]
Ahh ok I deleted my comment as I found if I removed some other code that uses this variable from that class, the class would compile so now I’m more confused 😂

**Comment by malware-dev** [comment]
Don't worry about it, that might very well be an ordering thing.

**Comment by AmandaCameron** [comment]
For the "The contextual keyword 'var' may only appear within a local variable declaration or in script code" message that this user (And I) am getting, it appears to be [this code](https://github.com/malforge/mdk2/blob/main/Source/Mdk.CommandLine/IngameScript/Pack/DefaultProcessors/SymbolRenamer.cs#L206-L212) being over-zelous and replacing a class variable's type with `var`, which is invalid. A minimal reproduction of this is below:

```csharp
using Sandbox.Game.EntityComponents;
using Sandbox.ModAPI.Ingame;
using Sandbox.ModAPI.Interfaces;
using SpaceEngineers.Game.ModAPI.Ingame;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using VRage;
using VRage.Collections;
using VRage.Game;
using VRage.Game.Components;
using VRage.Game.GUI.TextPanel;
using VRage.Game.ModAPI.Ingame;
using VRage.Game.ModAPI.Ingame.Utilities;
using VRage.Game.ObjectBuilders.Definitions;
using VRageMath;

namespace IngameScript
{
    partial class Program : MyGridProgram
    {

        public static System.Text.RegularExpressions.Regex navRegStr = new System.Text.RegularExpressions.Regex("^(\\{(\\S+)\\}){0,1}(\\[(\\S+)\\]){0,1}(.+)$");

        public void Main(string argument, UpdateType updateSource)
        {
            Echo($"Matches? {navRegStr.IsMatch(argument)}");
        }
    }
}
```

**Comment by malware-dev** [comment]
Thank you.

**Comment by malware-dev** [comment]
Fixed with 2.0.14
